packages/txservice/src/dispatch.ts

```diff
@@ -0,0 +1,195 @@
+  }
+
+  // Move latest pending to completed.
+  // @returns boolean whether operation was successful, or if there was no pending tx to shift.
+  public async shift(id: string): Promise<boolean> {
+    // TODO: Trim the validated stack when it gets "too long" to save memory space.
+    return await this.accessQueue.add((): boolean => {
+      const transaction = this.pending.find((transaction) => transaction.id === id);
+      if (transaction == null) {
+        return false;
+      }
+      this.validated.push(transaction);
+      return true;
+    });
+  }
+
+  /**
+   * Get the current nonce value.
+   *
+   * @remarks
+   * Caller should still be prepared to get the incorrect nonce back. For instance,
+   * if the provider that just handled our sent tx has suddenly gone offline, this
+   * method may give the wrong nonce. This can be solved by making additional calls to
+   * submit the tx.
+   *
+   * @returns A number value for the current nonce.
+   */
+  public async getNonce(): Promise<number | undefined> {
+    // Update nonce value to greatest of all nonce values retrieved.
+    return (await this.getLastTx())?.nonce;
+  }
+
+  private async getLastTx(): Promise<Transaction | null> {
+    return await this.accessQueue.add((): Transaction | null => {
+      if (this.pending.length > 0) {
+        return this.pending[this.pending.length - 1];
+      } else if (this.validated.length > 0) {
+        return this.validated[this.validated.length - 1];
+      }
+      return null;
+    });
+  }
+}
+
+// TODO: We may want to remodel dispatch to either extend ChainRpcProvider or just move all the ChainRpcProvider functionality over here.
+// For now, a sep. class for simplicity of introducing new functionality.
+/**
+ * @classdesc Dispatches and monitors transaction queue.
+ */
+export class TransactionDispatch {
+  public readonly chainId: number;
+
+  private readonly submitQueue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
+  // A queue for execution of transaction validation. When it completes, we'll shift the transaction from
+  // pending stack -> validated stack in the buffer.
+  private readonly validateQueue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
+  private readonly buffer: TransactionBuffer = new TransactionBuffer();
+
+  /**
+   * Centralized transaction dispatch management class.
+   *
+   * @param logger The pino.BaseLogger instance we use for logging.
+   * @param provider The ChainRpcProvider instance we use for interfacing with the chain.
+   */
+  constructor(private readonly logger: BaseLogger, private readonly provider: ChainRpcProvider) {
+    this.chainId = this.provider.chainId;
+  }
+
+  public async submit(minTx: WriteTransaction): Promise<Transaction> {
+    const method = this.submit.name;
+    // get gas estimate
+    let gasLimit: BigNumber;
+    let result = await this.provider.estimateGas(minTx);
+    if (result.isErr()) {
+      if (result.error instanceof TransactionReverted) {
```

Comment:
`instanceof` can have unpredictable behavior, you'll need to use a different condition here

Reply:
this was in original code so uh. dang that's dangerous. we should purge that from txservice entirely then, my b